### PR TITLE
[PROGRAMMES-6880] Structured data for recipes

### PR DESCRIPTION
**Pull request by tomwinnington**
Ticket:
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-6880

Link:
http://sandbox.bbc.co.uk/programmes/b013pqnm/recipes

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Controller/Recipes/RecipesController.php [hunk 1]
```diff
@@ -2,6 +2,7 @@
 namespace App\Controller\Recipes;
 
 use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use App\Controller\Helpers\StructuredDataHelper;
```
Comment (benolot) [inline src/Controller/Recipes/RecipesController.php]: Is this used?

File: src/Controller/Recipes/AbstractRecipesController.php [hunk 2]
```diff
@@ -34,17 +37,34 @@ public function __invoke(Programme $programme, RecipesService $recipesService)
...
     }
 
+    private function getSchema(StructuredDataHelper $structuredDataHelper, array $recipes, Programme $programme)
+    {
+        $recipeArray = [];
+        foreach ($recipes as $recipe) {
+            $recipeArray[] = $structuredDataHelper->getSchemaForRecipe($recipe);
+        }
+        if ($programme){
+            $schema = $structuredDataHelper->getSchemaForProgrammeContainer($programme);
```
Comment (benolot) [inline src/Controller/Recipes/AbstractRecipesController.php]: A programme isn't always a programme container (Which this method expects) - this is the cause of the errors in the CI test.
I suspect you need to add a method like getSchemaForProgramme() (If one doesn't exist already) that checks the object type and building the relevant schema.

File: src/Controller/Helpers/SchemaHelper.php [hunk 3]
```diff
@@ -218,4 +219,34 @@ public function buildSchemaForContributor(Contribution $contribution): array
...
+        $chef = $recipe->getChef();
+        $schema = [
+            '@type' => 'Recipe',
+            'url' => $recipe->getUrl(),
+            'name' => $recipe->getTitle(),
+            'about' => $recipe->getDescription(),
+
+        ];
+
+        if($recipe->getImage()){
```
Comment (benolot) [inline src/Controller/Helpers/SchemaHelper.php]:
```suggestion
        if ($recipe->getImage()) {
```

File: src/Controller/Helpers/SchemaHelper.php [hunk 4]
```diff
@@ -218,4 +219,34 @@ public function buildSchemaForContributor(Contribution $contribution): array
...
+            $schema['image'] = $recipe->getImage()->getUrl('480');
+        }
+
+        if($chef){
```
Comment (benolot) [inline src/Controller/Helpers/SchemaHelper.php]:
```suggestion
        if ($chef) {
```

File: src/Controller/Helpers/StructuredDataHelper.php [hunk 5]
```diff
@@ -74,6 +76,23 @@ public function getSchemaForEpisode(ProgrammeItem $programmeItem, bool $includeP
...
 
+    public function getSchemaForCoreEntity(CoreEntity $programme): array
+    {
+        if ($programme instanceof Episode) {
+            return $this->schemaHelper->getSchemaForEpisode($programme);
+        }
+        if ($programme instanceof Clip) {
+            return $this->buildSchemaForClip($programme);
+        }
+        if ($programme instanceof Series) {
```
Comment (BjornTwachtmann) [inline src/Controller/Helpers/StructuredDataHelper.php]: This is not needed

**Pull request merged**